### DISCUSSION TITLE
Example code on how to integrate AVM2 as an embedded Scripting Engine in a dummy C++ application

### DISCUSSION THREAD

**Discussion by al-sabr** [airsdk/Adobe-Runtime-Support#1449] [General]
Maybe you guys have some experience with this kind of specific needs. If you can direct us on how to be able to achieve that it would be really beneficial.

Best Regards

**Comment by ajwfrost** [comment]
Worth looking at the "avmshell" application as the best example for getting started. But, the AVM2 is not able to interpret ActionScript - there's no "eval" command, or anything like that (although there are some interesting workarounds for that which are themselves written in ActionScript... https://stackoverflow.com/questions/7971856/as3-evaluating-at-runtime-d-eval-vs-hurlant)

So you would still need to have a compiler turn the .as code into .abc or .swf, and then pass that into the application to be loaded by the avmplus engine.

**Comment by al-sabr** [comment]
> Worth looking at the "avmshell" application as the best example for getting started. But, the AVM2 is not able to interpret ActionScript - there's no "eval" command, or anything like that (although there are some interesting workarounds for that which are themselves written in ActionScript... https://stackoverflow.com/questions/7971856/as3-evaluating-at-runtime-d-eval-vs-hurlant)
> 
> So you would still need to have a compiler turn the .as code into .abc or .swf, and then pass that into the application to be loaded by the avmplus engine.

Could you provide some links for what you are referring to avmshell?

**Comment by ajwfrost** [comment]
The shell source code is at https://github.com/adobe/avmplus/tree/master/shell
Main loop is Shell::run within https://github.com/adobe/avmplus/blob/master/shell/avmshell.cpp
This calls the isolate 'run' method which would call the ShellIsolate::doRun() method (also in that file) which has 'evalCodeBlobs' call where it then calls 'evaluateScriptBuffer' on the core.

Essentially if you're going to handle just 'core' AS3 then you could get away with a very slim set of classes similar to the shell ones: but to actually make anything useful, you would then have a new set of AS3 APIs which you write as ActionScript classes with some 'native' calls (similar to Java/JNI). The build tools then create a set of stubs from these classes and you can add in the native (C++) implementations.

It can be a bit tricky (especially regarding memory management) - I'm curious as to the use case... it may be that it's something you could do just be extending the avmshell app itself, or even by just running an AIR application in command-line mode?
Also, check out the RedTamarin project which is another app that's using avmplus to provide the AS3 capabilities with a lot more focus on command-line and low-level API implementations: https://github.com/Corsaair/redtamarin

thanks

**Comment by al-sabr** [comment]
I was thinking about using the AVM2 with AS3 for basic smart contract compiled and put into some kind of Blockchain. Is there no example of a wrapper how to embed easily something or there is some kind of code already available to just use?

**Comment by al-sabr** [comment]
I have a found this repository... could you have a look an give your opinion about it?

https://github.com/drywolf/caspin

**Comment by ajwfrost** [comment]
That one looks interesting, but perhaps lacking some examples still of how to embed things. The test code perhaps shows how it can be done: execute a file, and then there are test functions that use the definitions created/loaded from within that file:
https://github.com/drywolf/caspin/tree/master/dev/tests/cpp_interface

So I guess if you go with that wrapper mechanism and then create all your logic in AS3 you can just do the same as they have there. The more complex part would be if you want to implement any of your AS3 functionality with C++ via the native calls..

As far as I'm aware though, there's no other example code other than avmshell and these third party projects..

thanks

**Reply by al-sabr** [reply]
İ think the basic features of AVM2 are perfect for a Blockchain VM where Smart Contract are compiled as .abc file.
